pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,297 @@
+	localProfileInterval := ctx.globalConfig.GlobalValueInt(types.ConfigInterval)
+	interval := time.Duration(localProfileInterval) * time.Second
+	max := float64(interval)
+	min := max * 0.3
+	ticker := flextimer.NewRangeTicker(time.Duration(min),
+		time.Duration(max))
+	// Return handle to caller
+	handleChannel <- ticker
+
+	wdName := agentName + "currentProfile"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	for {
+		select {
+		case <-ticker.C:
+			localProfileServer := getconfigCtx.localProfileServer
+			if localProfileServer != "" {
+				start := time.Now()
+				iteration++
+				localProfileURL, err := getLocalProfileURL(localProfileServer)
+				if err != nil {
+					log.Errorf("getLocalProfileURL: %s", err)
+				} else {
+					if localProfile, err := getLocalProfileConfig(localProfileURL, iteration, getconfigCtx); err != nil {
+						log.Errorf("localProfileTimerTask getLocalProfileConfig: %s", err)
+					} else {
+						if err := processProfileChange(getconfigCtx, localProfile); err != nil {
+							log.Errorf("localProfileTimerTask processProfileChange: %s", err)
+						}
+					}
+				}
+				ctx.ps.CheckMaxTimeTopic(wdName, "getLocalProfileConfig", start,
+					warningTime, errorTime)
+			}
+
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func parseLocalProfile(localProfileBytes []byte) (*profile.LocalProfile, error) {
+	var localProfile = &profile.LocalProfile{}
+	err := proto.Unmarshal(localProfileBytes, localProfile)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling failed: %v", err)
+	}
+	return localProfile, nil
+}
+
+func parseAndValidateLocalProfile(localProfileBytes []byte, getconfigCtx *getconfigContext) (*profile.LocalProfile, error) {
+	localProfile, err := parseLocalProfile(localProfileBytes)
+	if err != nil {
+		return nil, fmt.Errorf("parseAndValidateLocalProfile parseLocalProfile: %v", err)
+	}
+	if localProfile.GetServerToken() != getconfigCtx.profileServerToken {
+		// send something to ledmanager ??
+		return nil, fmt.Errorf("parseAndValidateLocalProfile: missamtch ServerToken for local profile server")
+	}
+	return localProfile, nil
+}
+
+// read saved local profile in case of particular reboot reason
+func readSavedLocalProfile(getconfigCtx *getconfigContext) (*profile.LocalProfile, error) {
+	localProfileMessage, err := readSavedProtoMessage(
+		getconfigCtx.zedagentCtx.globalConfig.GlobalValueInt(types.StaleConfigTime),
+		filepath.Join(checkpointDirname, savedLocalProfileFile), false)
+	if err != nil {
+		return nil, fmt.Errorf("lastlocalprofile: %v", err)
+	}
+	if localProfileMessage != nil {
+		log.Function("Using saved local profile")
+		return parseAndValidateLocalProfile(localProfileMessage, getconfigCtx)
+	}
+	return nil, errors.New("readSavedLocalProfile: empty message")
```

Comment:
This isn't an error but completely expected when there is no checkpointed proto file. Hence should return nil, nil to avoid spurious error logging.